### ISSUE TITLE
Question about the ratio during training

### ISSUE THREAD

**Issue by ViviTann** [yifan123/flow_grpo#104]
Thank you for the excellent work! I have a question regarding the ratio calculation during training.

In [line 643](https://github.com/yifan123/flow_grpo/blob/814021a2e6fcda9acfdda12e223638e6f17add4d/scripts/train_sd3.py?utm_source=chatgpt.com#L643)  and [line 879](https://github.com/yifan123/flow_grpo/blob/814021a2e6fcda9acfdda12e223638e6f17add4d/scripts/train_sd3.py?utm_source=chatgpt.com#L879) , `log_probs `are computed twice and then subtracted from each other: `(ratio = torch.exp(log_prob - sample["log_probs"][:, j]))`  
Could you explain the difference between these two `log_probs`?

Additionally, why is SDE inference first performed ` with torch.no_grad():` ([line 643](https://github.com/yifan123/flow_grpo/blob/814021a2e6fcda9acfdda12e223638e6f17add4d/scripts/train_sd3.py?utm_source=chatgpt.com#L643)), and then repeated with gradient computation enabled ([line 879](https://github.com/yifan123/flow_grpo/blob/814021a2e6fcda9acfdda12e223638e6f17add4d/scripts/train_sd3.py?utm_source=chatgpt.com#L879) )? Intuitively, I would expect the two corresponding `log_probs `to be the same, so I’d like to understand the reasoning behind this design.

**Comment by yifan123** [comment]
Thank you very much for the attention. This is an off-policy RL design. Line 643 computes the log\_probs when collecting data, while line 879 computes the log\_probs during training. The models used in these two places may differ — for example, if you collect data once but update twice, the model parameters will change after the first update. Therefore, line 643 corresponds to $P_\theta$, while line 879 corresponds to $P_{\theta_{\text{old}}}$.

<img width="1943" height="335" alt="Image" src="https://github.com/user-attachments/assets/f3b266ed-58d8-4058-b641-0a373c2424a4" />
